File: 2012w.ml
(*Problem 1*)
#use "helper.ml"

(*Problem 1:*)
let rec split l =
  let base =
  let fold_fn (i,l1,l2) elemt =
  let (_,l1,l2) = List.fold_left fold_fn base l in
  (l1,l2)


let rec merge l1 l2 =
  match (l1, l2) with
  | ([],l) -> l
  | (l,[]) -> l
  | 


let rec merge_sort l =


(*Problem 2*)
let replace s = implode (map (fun x -> if x = '-' then ' ' else x) (explode s))


(*Problem 3*)
let app l x = 

let [f1;f2] = app [(=);(<)] 2

let f11 x = false
let f12 x = 
let f13 x = 
let f21 x = 
let f22 x = 
let f23 x = 

File: final/2012w.pl
%4
%a
sorted([]).
sorted([_]).
sorted([X,Y|T]) :- X =< Y, sorted([Y|T]).


%b
delete(X,[X|T],T).
delete(X,[H|T],[H|NT]):-delete(X,T,NT).
permutation([], []).
permutation(L, [H|P]) :- delete(H,L,R), permutation(R,P).

sort2(L1,L2) :- permutation(L1,L2), sorted(L2),!.

%c
split([], [], []).
split([X], [X], []).
split([X|T],[X|L],R) :- split(T,R,L),!.

%d
merge([],L,L).
merge(L,[],L).
merge([H1|T1],[H2|T2],[H1|L]) :- H1=<H2, merge(T1,[H2|T2],L).
merge([H1|T1],[H2|T2],[H2|L]) :- H1>H2, merge([H1|T1],T2,L).

%e
merge_sort([],[]).
merge_sort([X],[X]).
merge_sort(L,S) :- split(L,P1,P2), merge_sort(P1,R1), merge_sort(P2,R2), merge(R1,R2,S),!.